api/pkg/s3/bucketsseput.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019 The OpenSDS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"crypto/md5"
+	"encoding/xml"
+	"fmt"
+	"github.com/opensds/multi-cloud/api/pkg/utils"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+	"github.com/opensds/multi-cloud/api/pkg/common"
+	. "github.com/opensds/multi-cloud/api/pkg/utils/constants"
+	"github.com/opensds/multi-cloud/s3/pkg/model"
+	s3 "github.com/opensds/multi-cloud/s3/proto"
+	log "github.com/sirupsen/logrus"
+)
+
+// Map from storage calss to tier
+var ClassAndTier1 map[string]int32
+var mutext1 sync.Mutex
+
+func (s *APIService) BucketSSEPut(request *restful.Request, response *restful.Response) {
+	bucketName := request.PathParameter("bucketName")
+	log.Infof("received request for PUT bucket SSE: %s", bucketName)
+
+	ctx := common.InitCtxWithAuthInfo(request)
+	bucket, err := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+	if err != nil {
+		log.Errorf("get bucket failed, err=%v\n", err)
+		response.WriteError(http.StatusInternalServerError, fmt.Errorf("bucket does not exist"))
+	}
+
+	body := ReadBody(request)
+	log.Infof("MD5 sum for body is %x", md5.Sum(body))
+
+	if body != nil {
+		sseConf := model.SSEConfiguration{}
+		errSSE := xml.Unmarshal(body, &sseConf)
+		if errSSE != nil {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		}
+
+		s3SSE  := &s3.ServerSideEncryption{
+			SseType:              "",
+			EncryptionKey:        nil,
+			InitilizationVector:  nil,
```

Comment:
@rhsakarpos it seems that the InitilizationVector is not used, but for encryption algo such as aes, initilization is needed.

Reply:
hi @hopkings2008 not really, IV is not needed for AES 256. If we use a custom salt, then we need an IV, this encrypts just fine. Thanks.

Reply:
@rhsakarpos because the core of encryption is to make the output message after encrypition more randomly, and currently, by using the iv for block-based cipher, the output message after encrypted will be more randomly. So, if we want to make the encryption more secure, it is better to use iv.  Just my opinion :)

Reply:
because the key generated in bucket sse put api is used to encrypt all the objects, so , iv should be use, pls refer to the link 'https://security.stackexchange.com/questions/35210/encrypting-using-aes-256-do-i-need-iv', thanks

Reply:
thanks @hopkings2008, will add post DAITO. Raised #750 to track this, assigned to me